primitives/blockchain/src/error.rs

```diff
@@ -35,112 +35,149 @@ pub enum ApplyExtrinsicFailed {
 	JustificationDecode,
+
 	/// Justification for header is correctly encoded, but invalid.
 	#[error("bad justification for header: {0}")]
 	BadJustification(String),
+
 	/// Not available on light client.
 	#[error("This method is not currently available when running in light client mode")]
 	NotAvailableOnLightClient,
+
 	/// Invalid remote CHT-based proof.
 	#[error("Remote node has responded with invalid header proof")]
 	InvalidCHTProof,
+
 	/// Remote fetch has been cancelled.
 	#[error("Remote data fetch has been cancelled")]
 	RemoteFetchCancelled,
+
 	/// Remote fetch has been failed.
 	#[error("Remote data fetch has been failed")]
 	RemoteFetchFailed,
+
 	/// Error decoding call result.
 	#[error("Error decoding call result of {0}")]
 	CallResultDecode(&'static str, #[source] CodecError),
+
 	/// Error converting a parameter between runtime and node.
 	#[error("Error converting `{0}` between runtime and node")]
 	RuntimeParamConversion(String),
+
 	/// Changes tries are not supported.
 	#[error("Changes tries are not supported by the runtime")]
 	ChangesTriesNotSupported,
+
 	/// Error reading changes tries configuration.
 	#[error("Error reading changes tries configuration")]
 	ErrorReadingChangesTriesConfig,
+
 	/// Key changes query has failed.
 	#[error("Failed to check changes proof: {0}")]
 	ChangesTrieAccessFailed(String),
+
 	/// Last finalized block not parent of current.
 	#[error("Did not finalize blocks in sequential order.")]
 	NonSequentialFinalization(String),
+
 	/// Safety violation: new best block not descendent of last finalized.
 	#[error("Potential long-range attack: block not in finalized chain.")]
 	NotInFinalizedChain,
+
 	/// Hash that is required for building CHT is missing.
 	#[error("Failed to get hash of block for building CHT")]
 	MissingHashRequiredForCHT,
+
 	/// Invalid calculated state root on block import.
 	#[error("Calculated state root does not match.")]
 	InvalidStateRoot,
+
 	/// Incomplete block import pipeline.
 	#[error("Incomplete block import pipeline.")]
 	IncompletePipeline,
+
+	/// Transaction pool initizliation is not complete.
 	#[error("Transaction pool not ready for block production.")]
 	TransactionPoolNotReady,
+
+	/// Database yielded an error.
 	#[error("Database")]
 	DatabaseError(#[from] sp_database::error::DatabaseError),
+
+	#[error("Failed to get header for hash {0}")]
+	MissingHashInHeader(String),
+
+	#[error("Failed to load the block weight for block {0}")]
+	MissingBlockWeightInHeader(String),
+
+	#[error("WASM override IO error")]
+	WasmOverrideIo(PathBuf, #[source] std::io::Error),
+
+	#[error("Overwriting WASM requires a directory where local \
+	WASM is stored. {0} is not a directory", .0.display())]
+	WasmOverrideNotADirectory(PathBuf),
+
+	#[error("Duplicate WASM Runtimes found: \n{}\n", .0.join("\n") )]
+	DuplicateWasmRuntime(Vec<String>),
+
 	/// A convenience variant for String
+	#[deprecated(note = "Introduce more typed error variants as needed")]
```

Comment:
And what should downstream projects do?

And either remove it or don't add a deprecation message.

Reply:
Define their own error type and forward it - and that's missing: `Foreign(Box<Error + Sync + Send+ 'static>)`